File: libs/jetpack/lib-arch/src/main/java/com/yyxnb/what/arch/viewmodel/BaseViewModel.java
package com.yyxnb.what.arch.viewmodel;


import androidx.annotation.NonNull;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.ViewModel;

/**
 * ================================================
 * 作    者：yyx
 * 版    本：1.0
 * 日    期：2020/11/21
 * 历    史：
 * 描    述：实现了LifecycleOwner接口，使其持有生命周期。可在 {@link BaseViewModel#onCreate()} 观察生命周期事件
 * ================================================
 */
public abstract class BaseViewModel extends ViewModel implements LifecycleOwner {

    private LifecycleOwner mLifecycleOwner;

    /**
     * 观察生命周期事件
     */
    protected abstract void onCreate();

    @SuppressWarnings("unchecked")
    public <T extends BaseViewModel> T attachLifecycleOwner(LifecycleOwner lifecycleOwner) {
        T currentModel = (T) this;

        if (null != mLifecycleOwner) {
            return currentModel;
        }

        this.mLifecycleOwner = lifecycleOwner;
        onCreate();
        return currentModel;
    }


    @NonNull
    @Override
    public Lifecycle getLifecycle() {
        return mLifecycleOwner.getLifecycle();
    }

    /**
     * onDestroy 调用时调用
     */
    @Override
    protected void onCleared() {
    }
}

File: arch/src/main/java/com/yyxnb/arch/base/mvvm/BaseViewModel.kt
package com.yyxnb.arch.base.mvvm

import android.arch.lifecycle.DefaultLifecycleObserver
import android.arch.lifecycle.LifecycleOwner
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import android.support.annotation.CallSuper
import com.yyxnb.arch.ext.tryCatch
import kotlinx.coroutines.*


/**
 * 逻辑处理
 *
 * 负责数据处理和View层与Model层的交互。
 * ViewModel通过数据仓库Repository获取数据来源，处理来自View的事件命令，同时更新数据。
 * @author : yyx
 * @date ：2018/6/13
 */
abstract class BaseViewModel : ViewModel(), DefaultLifecycleObserver {

    val msg: MutableLiveData<String> = MutableLiveData()

    open val mScope: CoroutineScope by lazy {
        CoroutineScope(SupervisorJob() + Dispatchers.Main)
    }

    fun launchUI(block: suspend CoroutineScope.() -> Unit) {
        mScope.launch {
            tryCatch({
                block()
            }, {
                throw it
            })
        }
    }

    suspend fun <T> launchOnIO(block: suspend CoroutineScope.() -> T) {
        withContext(Dispatchers.IO) {
            block
        }
    }

    override fun onCreate(owner: LifecycleOwner) {
        super.onCreate(owner)
    }

    override fun onDestroy(owner: LifecycleOwner) {
        super.onDestroy(owner)
    }

    @CallSuper
    override fun onCleared() {
        super.onCleared()
        mScope.cancel()
    }
}
